// inquirer-table-input/index.js
const chalk = require("chalk");
const cliCursor = require("cli-cursor");
const inquirer = require("inquirer");
const Base = require("inquirer/lib/prompts/base");
const Choices = require("inquirer/lib/objects/choices");
const observe = require("inquirer/lib/utils/events");
const Paginator = require("inquirer/lib/utils/paginator");
const Table = require("cli-table");
const { map, takeUntil } = require("rxjs/operators");

class TableInput extends Base {
  /**
   * Initialise the prompt
   *
   * @param  {Object} questions
   * @param  {Object} rl
   * @param  {Object} answers
   */
  constructor(questions, rl, answers) {
    super(questions, rl, answers);

    // private variables
    this.showInfo = true;
    this.columns = new Choices(this.opt.columns, []);
    this.verticalPointer = 0;
    this.nextEscapeWillClose = false;

    // parameters
    this.opt.escapeMessage =
      this.opt.escapeMessage || chalk.red("Press ESC again to exit!");
    this.opt.confirmMessage =
      this.opt.confirmMessage || chalk.green("Press ENTER again to confirm!");
    this.opt.hideInfoWhenKeyPressed = this.opt.hideInfoWhenKeyPressed || false;
    this.opt.infoMessage = this.opt.infoMessage || "";
    this.opt.decimalPoint = this.opt.decimalPoint || ".";
    this.opt.decimalPlaces = this.opt.decimalPlaces || 2;
    this.opt.freezeColumns = this.opt.freezeColumns || 0;
    this.selectedColor = this.opt.selectedColor || chalk.yellow;
    this.editableColor = this.opt.editableColor || chalk.bgYellow;
    this.editingColor = this.opt.editingColor || chalk.bgBlue;
    this.horizontalPointer = this.opt.freezeColumns;
    this.rows = new Choices(this.opt.rows, []);
    this.values = this.columns.filter(() => true).map(() => undefined);

    this.pageSize = this.opt.pageSize || 5;
  }

  /**
   * Start the inquirer session
   *
   * @param  {Function} callback
   * @return {TablePrompt}
   */
  _run(callback) {
    this.done = callback;

    const events = observe(this.rl);
    const validation = this.handleSubmitEvents(
      events.line.pipe(map(this.getCurrentValue.bind(this)))
    );
    validation.success.forEach(this.onEnd.bind(this));
    validation.error.forEach(this.onError.bind(this));

    events.keypress.forEach(({ key }) => {
      if (this.opt.hideInfoWhenKeyPressed) this.showInfo = false;
      if (key.name !== "escape") this.nextEscapeWillClose = false;
      this.nextEnterWillConfirm = false;

      if (!this.editingMode) {
        switch (key.name) {
          case "down":
            return this.onDownKey();

          case "up":
            return this.onUpKey();

          case "left":
            return this.onLeftKey();

          case "right":
            return this.onRightKey();

          case "escape":
            if (!this.nextEscapeWillClose) {
              this.nextEscapeWillClose = true;
              return this.render(this.opt.escapeMessage);
            } else {
              this.render();
              return this.onEnd(false);
            }

          default:
            if (this.columns.get(this.horizontalPointer).editable) {
              this.updateEditing();
              this.onEditPress(key);
            }
        }
      } else {
        return this.onEditPress(key);
      }
    });

    cliCursor.hide();
    this.render();

    return this;
  }

  getCurrentValue() {
    const currentValue = [];

    this.rows.forEach((row, rowIndex) => {
      currentValue.push(this.values[rowIndex]);
    });

    return currentValue;
  }

  onUpKey() {
    this.verticalPointer = this.verticalPointer > 0 ? this.verticalPointer - 1 : this.verticalPointer;
    this.render();
  }

  onDownKey() {
    const length = this.rows.realLength;

    this.verticalPointer = this.verticalPointer < length - 1 ? this.verticalPointer + 1 : this.verticalPointer;
    this.render();
  }

  onLeftKey() {
    const length = this.columns.realLength - 1;

    this.horizontalPointer =
      this.horizontalPointer > this.opt.freezeColumns
        ? this.horizontalPointer - 1
        : length;
    this.render();
  }

  onRightKey() {
    const length = this.columns.realLength;

    this.horizontalPointer =
      this.horizontalPointer < length - 1
        ? this.horizontalPointer + 1
        : this.opt.freezeColumns;
    this.render();
  }

  onEnd(state) {
    this.status = "answered";
    this.screen.done();
    cliCursor.show();

    if (state) {
      return this.done({
        state,
        result: this.rows.choices.map(row => {
          const rowData = {};
          this.columns.choices.forEach((column, index) => {
            rowData[column.value] = row[index]
              .toString()
              .replace(/\u001b\[[0-9;]*m/g, "");
          });
          return rowData;
        })
      });
    } else {
      return this.done({
        state
      });
    }
  }

  formatCell() {
    if (this.editingType === "number") {
      this.rows.choices[this.verticalPointer][this.horizontalPointer] = parseInt(
        this.rows.choices[this.verticalPointer][this.horizontalPointer] || 0
      );
    }

    if (this.editingType === "decimal") {
      this.rows.choices[this.verticalPointer][this.horizontalPointer] = parseFloat(
        this.rows.choices[this.verticalPointer][this.horizontalPointer] || 0
      ).toFixed(this.opt.decimalPlaces);
    }

    if (this.editingType === "boolean") {
      const value = String(
        this.rows.choices[this.verticalPointer][this.horizontalPointer]
      ).toLowerCase();

      this.rows.choices[this.verticalPointer][this.horizontalPointer] =
        value === "true";
    }

    if (this.editingType === "choice") {
      const choices = this.columns.get(this.horizontalPointer).choices;
      let current = this.rows.choices[this.verticalPointer][this.horizontalPointer];

      if (!choices.includes(current)) {
        this.rows.choices[this.verticalPointer][this.horizontalPointer] = choices[0];
      }
    }
  }

  onError(state) {
    if (this.editingMode) {
      this.editingMode = false;
      this.formatCell();
      this.render();
    } else {
      if (!this.nextEnterWillConfirm) {
        this.nextEnterWillConfirm = true;
        return this.render(this.opt.confirmMessage);
      } else {
        this.render();
        return this.onEnd(true);
      }
    }
  }

  onEditPress(key) {
    this.editingType = this.columns
      .get(this.horizontalPointer)
      .editable.toLowerCase();
    this.isNumber = /^[0-9]$/.test(key.name);
    this.isText = /^[\x20-\x7E]$/.test(key.sequence);
    this.isDecimal = /^[0-9\,\.]$/.test(key.sequence);
    this.isLeftRight = /^(left|right)$/.test(key.name);

    // Exemplo: Adicione o caractere pressionado à sua escolha atual
    let value = this.rows.choices[this.verticalPointer][this.horizontalPointer];

    if (this.firstTimeEditingMode) {
      this.valueBeforeEditing = value;
      this.firstTimeEditingMode = false;
      value = "";
    }

    switch (key.name) {
      case "escape": {
        this.rows.choices[this.verticalPointer][
          this.horizontalPointer
        ] = this.valueBeforeEditing;
        this.render();
        return this.updateEditing();
      }
      case "delete": {
        if (this.editingType === "boolean" || this.editingType === "choice") {
          return false;
        }
        this.rows.choices[this.verticalPointer][this.horizontalPointer] = "";
        this.formatCell();
        return this.updateEditing();
      }
      case "backspace": {
        if (this.editingType === "boolean" || this.editingType === "choice") {
          return false;
        }
        let value = this.rows.choices[this.verticalPointer][
          this.horizontalPointer
        ].toString();
        this.rows.choices[this.verticalPointer][this.horizontalPointer] = value.slice(
          0,
          -1
        );

        value = this.rows.choices[this.verticalPointer][
          this.horizontalPointer
        ].toString();
        if (value.length === 0) {
          this.formatCell();
          this.updateEditing();
        }

        this.render();
      }
      default: {
        if (this.editingType === "number") {
          if (this.isNumber) {
            this.rows.choices[this.verticalPointer][this.horizontalPointer] =
              value + key.name;
          }
          this.render();
        }
        if (this.editingType === "text") {
          if (this.isText) {
            this.rows.choices[this.verticalPointer][this.horizontalPointer] =
              value + key.sequence;
          }
          this.render();
        }
        if (this.editingType === "decimal") {
          if (this.isDecimal) {
            const existDecimalPoint =
              value.indexOf(".") >= 0 || value.indexOf(",") >= 0;
            const decimalPointPressed =
              key.sequence.indexOf(".") >= 0 || key.sequence.indexOf(",") >= 0;

            const keyAccepted =
              !decimalPointPressed || (!existDecimalPoint && decimalPointPressed);

            if (keyAccepted) {
              this.rows.choices[this.verticalPointer][this.horizontalPointer] =
                value +
                key.sequence
                  .replace(".", this.opt.decimalPoint)
                  .replace(",", this.opt.decimalPoint);

            }
          }
          this.render();
        }
        if (this.editingType === "boolean") {
          if (this.isLeftRight) {
            const current = this.rows.choices[this.verticalPointer][this.horizontalPointer];
            const currentBool = String(current).toLowerCase() === "true";
            this.rows.choices[this.verticalPointer][this.horizontalPointer] = (!currentBool).toString();
          }
          this.render();
        }

        if (this.editingType === "choice") {
          if (this.isLeftRight) {
            const choices = this.columns.get(this.horizontalPointer).choices;
            const currentValue = this.rows.choices[this.verticalPointer][this.horizontalPointer];
            const currentIndex = choices.indexOf(currentValue);

            if (currentIndex !== -1) {
              const direction = key.name === "left" ? -1 : 1;
              const newIndex = (currentIndex + direction + choices.length) % choices.length;

              this.rows.choices[this.verticalPointer][this.horizontalPointer] = choices[newIndex];
            }
          }
          this.render();
        }
        return false;
      }
    }
  }

  updateEditing() {
    const isEditable = this.columns.get(this.horizontalPointer).editable;

    if (isEditable) {
      this.firstTimeEditingMode = true;
      this.editingMode = !this.editingMode;
      this.render();
    }
  }


  paginate() {
    const middleOfPage = Math.floor(this.pageSize / 2);
    const firstIndex = Math.max(0, this.verticalPointer - middleOfPage);
    const lastIndex = Math.min(
      firstIndex + this.pageSize - 1,
      this.rows.realLength - 1
    );
    const lastPageOffset = this.pageSize - 1 - lastIndex + firstIndex;

    return [Math.max(0, firstIndex - lastPageOffset), lastIndex];
  }

  render(message) {
    let content = this.getQuestion();
    let bottomContent = "";

    if (this.showInfo) {
      content += this.opt.infoMessage;
    }

    const [firstIndex, lastIndex] = this.paginate();
    const table = new Table({
      head: this.columns.pluck("name").map(name => name)
    });

    this.rows.forEach((row, rowIndex) => {
      if (rowIndex < firstIndex || rowIndex > lastIndex) return;

      const columnValues = [];

      this.columns.forEach((column, columnIndex) => {
        const editable =
          (column.editable &&
            column.editable.toString().toLowerCase() !== "none") ||
          false;

        const isSelected =
          this.verticalPointer === rowIndex && this.horizontalPointer === columnIndex;

        const cellValue = this.rows.realChoices[rowIndex][columnIndex];

        const value = editable ? ` ${cellValue} ` : cellValue;

        if (this.editingMode) {
          columnValues.push(
            isSelected
              ? editable
                ? this.editingColor(value)
                : this.selectedColor(value)
              : value
          );
        } else {
          columnValues.push(
            isSelected
              ? editable
                ? this.editableColor(value)
                : this.selectedColor(value)
              : value
          );
        }
      });

      table.push(columnValues);
    });

    content += "\n\n" + table.toString();

    if (message) {
      bottomContent = message;
    }

    this.screen.render(content, bottomContent);
  }
}

module.exports = TableInput;
